File: JDK11/opentraceability/src/main/java/opentraceability/utility/enumutil.java
package opentraceability.utility;

import opentraceability.OTLogger;
import java.lang.annotation.Annotation;
import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.List;

public class EnumUtil {
    public static <T extends Enum<T>> String GetEnumDescription(T value) {
        try {
            Field field = value.getClass().getField(value.name());
            opentraceability.utility.attributes.Description[] displayAnnotation = field.getAnnotationsByType(opentraceability.utility.attributes.Description.class);
            return displayAnnotation.length > 0 ? displayAnnotation[0].Description() : null;
        } catch (Exception ex) {
            OTLogger.error(ex);
            throw ex;
        }
    }

    public static <T extends Enum<T>> String GetEnumDisplayName(T value) {
        try {
            Field field = value.getClass().getField(value.name());
            opentraceability.utility.attributes.Display[] displayAnnotation = field.getAnnotationsByType(opentraceability.utility.attributes.Display.class);
            return displayAnnotation.length > 0 ? displayAnnotation[0].Name() : null;
        } catch (Exception ex) {
            OTLogger.error(ex);
            throw ex;
        }
    }

    public static <T extends Enum<T>, A extends Annotation> List<A> GetEnumAttributes(T value) {
        try {
            Field field = value.getClass().getField(value.name());
            A[] annotations = field.getAnnotationsByType((Class<A>) Annotation.class);
            List<A> annotationList = new ArrayList<>();
            for (A annotation : annotations) {
                annotationList.add(annotation);
            }
            return annotationList;
        } catch (Exception ex) {
            OTLogger.error(ex);
            throw ex;
        }
    }

    public static <T extends Enum<T>> List<T> GetValues() {
        return List.of(Enum.<T>values());
    }
}

File: Kotlin/OpenTraceability/src/main/kotlin/opentraceability/utility/enumutil.kt
package opentraceability.utility

import opentraceability.OTLogger
import java.util.*

object EnumUtil {
    inline fun <reified T : Enum<T>> GetEnumDescription(value: T): String? {
        try {
            val field = value::class.java.getField(value.name)
            val displayAnnotation = field.getAnnotationsByType(opentraceability.utility.attributes.Description::class.java)
            return displayAnnotation.firstOrNull()?.Description
        } catch (ex: Exception) {
            opentraceability.OTLogger.error(ex)
            throw ex
        }
    }

    inline fun <reified T : Enum<T>> GetEnumDisplayName(value: T): String? {
        try {
            val field = value::class.java.getField(value.name)
            val displayAnnotation = field.getAnnotationsByType(opentraceability.utility.attributes.Display::class.java)
            return displayAnnotation.firstOrNull()?.Name
        } catch (ex: Exception) {
            opentraceability.OTLogger.error(ex)
            throw ex
        }
    }

    inline fun <T : Enum<T>, reified A : Annotation> GetEnumAttributes(value: T): MutableList<A> {
        try {
            val field = value::class.java.getField(value.name)
            val annotations = field.getAnnotationsByType(A::class.java)
            return annotations.toMutableList()
        } catch (ex: Exception) {
            opentraceability.OTLogger.error(ex)
            throw ex
        }
    }

    inline fun <reified T : Enum<T>> GetValues(): MutableList<T> {
        return enumValues<T>().toMutableList()
    }
}
